src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchShareLogDAO.scala

```diff
@@ -0,0 +1,163 @@
+    shareType match {
+      case Some(typeOfShare) =>
+        boolQuery().filter(
+          boolQuery()
+            .must(userShares(userId))
+            .must(termQuery("shareType", typeOfShare.toString)))
+      case _ => userShares(userId)
+    }
+  }
+
+}
+
+/**
+  * DAO that uses ElasticSearch to log and get records of shares.
+  *
+  * @param client      The ElasticSearch client
+  * @param indexName   The name of the target share log index in ElasticSearch
+  */
+class ElasticSearchShareLogDAO(client: TransportClient, indexName: String, refreshMode: RefreshPolicy = RefreshPolicy.NONE)
+  extends ShareLogDAO with ElasticSearchDAOSupport with ShareQueries {
+
+  lazy private final val datatype = "sharelog"
+
+  init // checks for the presence of the index
+
+  /**
+    * Logs a record of a user sharing a workspace, group, or method with a user.
+    *
+    * @param userId     The workbench user id
+    * @param sharee     The email of the user being shared with
+    * @param shareType  The type (workspace, group, or method) see `ShareLog`
+    * @return           The record of the share
+    */
+  override def logShare(userId: String, sharee: String, shareType: ShareType.Value): Share = {
+    val share = Share(userId, sharee, shareType, Some(Instant.now))
+    val id = generateId(share)
+    val insert = client
+      .prepareIndex(indexName, datatype, id)
+      .setSource(share.toJson.compactPrint, XContentType.JSON)
+      .setRefreshPolicy(refreshMode)
+
+    executeESRequest[IndexRequest, IndexResponse, IndexRequestBuilder](insert)
+    share
+  }
+
+  /**
+    * Logs records of a user sharing a workspace, group, or method with users.
+    * todo could be a batch query
+    *
+    * @param userId The workbench user id
+    * @param sharees The emails of the users being shared with
+    * @param shareType The type (workspace, group, or method) see `ShareLog`
+    * @return The records of the shares - see `ShareLog.Share`
+    */
+  override def logShares(userId: String, sharees: Seq[String], shareType: ShareType.Value): Seq[Share] = {
+    sharees map { sharee => logShare(userId, sharee, shareType) }
+  }
+
+  /**
+    * Gets a share by the ID, a `MurmurHash3` of `userId` + `sharee` + `shareType`
+    *
+    * @param share The share to get
+    * @return A record of the share
+    */
+  override def getShare(share: Share): Share = {
+    val id = generateId(share)
+    val getSharesQuery = client.prepareGet(indexName, datatype, id)
+    Try(executeESRequest[GetRequest, GetResponse, GetRequestBuilder](getSharesQuery)) match {
+      case Success(get) if get.isExists => get.getSourceAsString.parseJson.convertTo[Share]
+      case Success(_) => throw new FireCloudException(s"share not found")
+      case Failure(f) => throw new FireCloudException(s"error getting share for $share: ${f.getMessage}")
+    }
+  }
+
+  /**
+    * Gets all shares that have been logged for a workbench user which fall under the
```

Comment:
this will get "all" up to a limit of 100. We won't hit that too soon, but it may cause a problem in the long run. Mention that in the comment!

Reply:
Would it be worth adding a `log.debug` when 100 items are returned?

Reply:
our default log level is `info` so we'd never see it! https://github.com/broadinstitute/firecloud-orchestration/blob/develop/src/main/resources/logback.xml#L22

I think a `warn` is good here. That'll trigger Sentry - but we should be bugged about it, because if we reach that point we need to update code/rethink approach.